### PR TITLE
Multi-dag LRU caching in Light, support per-instance caches

### PR DESCRIPTION
**Pull request by karalabe**
Ethash in its current state contains two issues, the combination of which makes epoch transitions very expensive:

 * A single verification DAG cache is used ever, always being replaced whenever a different epoch is requested. This is problematic during epoch transitions since the block is in the new epoch while the uncle is in the old. Similarly this causes trouble when processing blocks in parallel, since they trash the cache.
 * A single global light DAG was shared between all instances of Ethash. This was done to allow reusing underlying caches between test runs. Although this is beneficial in some scenarios, it will cause trouble later if a process uses 2+ instances of Ethash (e.g. connected simultaneously to main net and test net), as it will introduce the same cache trashing.

This PR fixes both of these scenarios, while maintaining the existing feature set:

 * The `Light` structure was expanded with a simple LRU cache implementation, that accumulates a maximum of 3 verification DAGs before starting to evict old ones (least recently retrieved). Although during normal operation 2 DAGs should suffice, the reason the 3rd one was inserted is to avoid cache trashing when older blocks also circulate the network.
 * Regarding the shared light structure: during normal mode of operation the user would expect `Ethash` instances to be independent of each other, so the plain `ethash.New` method was fixed to use its own `Light` struct. However, to keep the ability to reuse caches during tests, I've added a `NewShared` method to ethash, which is essentially the old `New`, reusing the same shared Light cache.

### CODE REVIEW

File: ethash.go [hunk 1]
```diff
@@ -104,14 +105,25 @@ func freeCache(cache *cache) {
...
+// Light implements the Verify half of the proof of work. It uses a few small
+// in-memory caches to verify the nonces found by Full.
 type Light struct {
-	test    bool       // if set use a smaller cache size
-	mu      sync.Mutex // protects current
-	current *cache     // last cache which was generated.
-	// TODO: keep multiple caches.
+	test  bool              // if set use a smaller cache size
+	mu    sync.Mutex        // protects the per-epoch map of DAGs
+	dags  map[uint64]*cache // currently cached verification DAGs
```
Comment (fjl) [inline ethash.go]: Can we rename this to "caches"? Ethash terminology distinguishes cache (for verification) and DAG (for mining).

**Comment by karalabe** [comment]
@fjl PTAL

**Comment by fjl** [comment]
:+1:

**Comment by obscuren** [comment]
:+1:

**Pull request merged**